compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -578,6 +580,183 @@ static int64_t getMaxVectorTileSize(int64_t numElem, int64_t tileSize,
   return 1;
 }
 
+static bool isInnerMostDimThatMapIsFunctionOf(AffineMap map, int dim) {
+  if (!map.isFunctionOfDim(dim)) {
+    return false;
+  }
+  for (int d = dim + 1; d < map.getNumDims(); ++d) {
+    if (map.isFunctionOfDim(d)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static int sum(ArrayRef<int64_t> a) {
+  int s = 0;
+  for (int i : a) {
+    s += i;
+  }
+  return s;
+}
+
+// Clamps in-place `vecTileSizes`, ensuring that the resulting vector tile sizes
+// for each opearand of `op` satisfy two requirements:
+// 1. No resulting operand tile size exceeds `eachOperandMaxTileBits`.
+// 2. The sum of all resulting operand tile size does not exceed
+// `allOperandsMaxTileBits`.
+static void limitVectorTileSizes(linalg::LinalgOp op,
+                                 SmallVectorImpl<int64_t> &vecTileSizes,
+                                 int64_t eachOperandMaxTileBits,
+                                 int64_t allOperandsMaxTileBits) {
+  int numLoops = op.getNumLoops();
+  assert(numLoops == vecTileSizes.size());
+  auto indexingMaps = op.getIndexingMapsArray();
+  auto operandTypes = op->getOperandTypes();
+  int numOperands = operandTypes.size();
+
+  SmallVector<int64_t> operandElemBits(numOperands);
+  for (int i : llvm::seq<int>(0, numOperands)) {
+    operandElemBits[i] =
+        cast<ShapedType>(operandTypes[i]).getElementTypeBitWidth();
+  }
+
+  // For each operand, we track how big the tile is going to be based on the
+  // dimensions that we have already accounted for. Here we initialize this to
+  // just elemBitWidth, having not yet accounted for any array dimension.
+  SmallVector<int64_t> tileBits(numOperands);
+  for (int i : llvm::seq<int>(0, numOperands)) {
+    tileBits[i] = operandElemBits[i];
```

Comment:
Using `llvm::enumerate` is cleaner when you need both index and value. So you can do something like:

```cpp
for (auto [i, val] : llvm::enumerat(operandElemBits)) {
  tileBits[i] = val;
  // ...
}
```

or something like:

```cpp
for (auto bits : operandElemBits) {
  tileBits.push_back(bits);
  if (tileBits.back() > ... ) {
    // ...
  }
}
```

SmallVector is a vector optimized for small arrays. This optimization comes from not performing heap allocations for a limited number of elements. So I think we don't need to reserve the space for the vector in this case.